### ISSUE TITLE
Add ability to disable certain CRDs on installation.

### ISSUE THREAD

**Issue by mvaal** [crossplane/crossplane#2869]
Edit from @negz:

```
I0328 13:34:48.091330 25599 request.go:665] Waited for 1.033772408s due to client-side throttling, not priority and fairness, request: GET:https://xyz.com/apis/pkg.crossplane.io/v1alpha1?timeout=32s
```

**If you're finding this issue because you're seeing messages like the above when you run kubectl please make sure you're using the latest version of kubectl - i.e. 1.25.0 or above.** Similarly, other tools built against client-go 0.25.0 or above should not experience these issues that are due to overly aggressive client-side rate limits.

Original issue follows below.

---

### What problem are you facing?
We want to limit which resources are installed on our cluster so that we can restrict which resources can be used in AWS by our cluster's users.
In addition to this, we are having throttling issues due to the large amount of CRDs, being able to turn of 90% of the the unused CRDs would make the throttling issues go away.

### Related issues

- [x] #2895 
- [x] https://github.com/crossplane/docs/issues/336

**Comment by muvaf** [comment]
@mvaal I have transferred the issue to core crossplane.

@liubog2008 opened https://github.com/crossplane/provider-aws/pull/1076 to do this for controllers, I think consolidating the whole discussion in this issue would be more helpful. I've [answered](https://github.com/crossplane/provider-aws/pull/1076#issuecomment-1017162860) why we didn't do it for Jet-based providers which have way more CRDs.

It seems like the main concerns are as following:
* Ability to limit capabilities of users in the cluster.
  * This can be addressed with RBAC for most cases but in cases where you don't want `Composition` authors be able to use certain managed resource types, there isn't much you can do other than using OpenPolicyAgent because the creator of managed resources resulting from `Composition` is Crossplane pod, which has permissions over all CRDs packages bring.
* Throttling of client requests.
  * Even if throttling wasn't there, the number of requests discovery client makes increases linearly with number of CRDs which makes non-cached queries run longer.

@mvaal @liubog2008 could you add other concerns if you have any? I'd like to have all the feedback in one place so that when we reconsider, we have more data to depend on.

cc @negz @hasheddan

**Comment by mvaal** [comment]
You pretty much nailed it. We are using OPA, but we would rather just not worry about the unused resources at all and add them when required.
As for the throttling issue, its mostly in reference to this: https://github.com/kubernetes/kubectl/issues/1126
They are already talking about crossplane in that thread as well.

**Comment by haarchri** [comment]
we have the same requirement from security department to not deploy all CRDs - only approved CRDs - which matches our IAM Policies ... otherwise we need to write a lot of documentation - because of that fact currently we build own providers and skip CRDs / Functions - an option to enable only a few CRDs could help

**Comment by negz** [comment]
Thanks for the feedback folks. It's helpful to understand the use cases here. To summarize, I'm seeing:

* **Performance degradation**. So far we've preferred trying to work with upstream to fix the performance issues rather than working around them in Crossplane. Disabling CRDs seems like it at best delays this issue, or reduces the amount of folks who will hit it. Presumably some people will actually want to use many type of custom resource.
* **Restricting allowed resources**. It sounds like the problem here is an interaction between RBAC and Composition. Our position has been that you should use RBAC to restrict who can do what, not avoid installing types completely, but as you point out anyone who has the ability to write a `Composition` has the ability to ask Crossplane to create any kind of resource.
* **Security policy**. The security department wants to approve specific CRDs.

@mvaal One option to manage which managed resources a `Composition` author can ask Crossplane to create would be to avoid using the Crossplane RBAC manager and instead manually curate an RBAC role for Crossplane. As @muvaf mentioned though, we tend to assume that the folks authoring/installing a `Composition` are the same folks installing Providers. Is this not true for your organization?

@haarchri Can you help me understand what kind of things your security department is auditing on a custom resource by custom resource level? Would RBAC restrictions help in your case?

**Comment by glaxman** [comment]
We started installing the aws terra jet provider using the following command, and it is taking too long to install all the cards. Moreover, we see a lot of client calls throttling. Is there a way to configure what CRDs we can install rather than installing all the CRDs?

**Comment by muvaf** [comment]
Cross reference to https://github.com/crossplane/crossplane/issues/2122

**Comment by ulucinar** [comment]
In the context of #2122, we had previously considered extending the `Providers.pkg` API and correspondingly the `kubectl crossplane install provider` command to support selective installation of CRDs and selectively starting reconcilers (starting only reconcilers for the enabled CRDs) with the following PR (and a bunch of other related PRs in the provider repos):
https://github.com/crossplane/crossplane/pull/2646

The proposed `Provider.pkg` API looked like as follows (more details are in the description of #2646):
```yaml
apiVersion: pkg.crossplane.io/v1
kind: Provider
metadata:
  name: crossplane-provider-tf-azure
spec:
  enabledAPIs:
  - Provider.*
  - virtual.*
  - lb
  - resource.azure.tf.crossplane.io/v1alpha1.ResourceGroup
  package: ulucinar/provider-tf-azure-arm64:build-83b9bfc0
```
The proposed default behavior of the new API (if no enabled APIs (CRDs) are defined) was to enable all, i.e., `.*`, to be backwards-compatible. Similarly, one can optionally specify the CRDs to be enabled (and their reconcilers to be started) using:
`kubectl crossplane install provider  --enable-api="virtual.* lb resource.azure.tf.crossplane.io/v1alpha1.ResourceGroup"

**Comment by negz** [comment]
@ulucinar regarding https://github.com/crossplane/crossplane/issues/2869#issuecomment-1108422111, I recall we had some technical concerns with this implementation, but I unfortunately can't find them written down anywhere.

One issue was how we'd handle cross-resource-references. For example you enable API group A and disable API group B, but API group A can reference API group B. I believe in https://github.com/crossplane-contrib/provider-jet-azure/pull/74 (the reference implementation at the provider level) disabled types were not being added to the `Scheme`, which I _think_ would cause a panic (maybe just an error?) in `client.Get` when an enabled API type tried to reference a disabled API type.

Things might be a little better if we kept all types added to the `Scheme` at the code level but simply didn't install the CRD into the API server. In that scenario I imagine references would just fail indefinitely with an error that the referred type didn't exist. In my opinion this would not be a great user experience, since it _appears_ possible to reference a thing that actually doesn't exist, but perhaps it would be livable.

I also vaguely recall there might have been some more complications in the package manager and package revisions around CRD ownership. @hasheddan does that sound familiar at all?

**Comment by carpenterm** [comment]
I just started looking at Crossplane and came across this issue as was suddenly experiencing throttling and wanted to cut down the number of CRDs to just those I want to use. It's great that the Crossplane team have been focusing on fixing performance issues upstream, thank you for those efforts. I feel those issues were important to fix for scenarios where people genuinely have 1000s of CRDs they want to use on a given cluster rather than enabling us to deploy huge numbers of objects that will never be used. It's totally fine to do deploy all of these by default but the user having control over what gets deployed would be so much better.

Does it stand to reason that retrieving a list of 1000s of resources using kubectl/k8s API is always going to be slower than a list of 10s of resources, no matter how much you fix upstream? So if a user ends up waiting longer and the majority of those resources are unnecessary then this is very frustrating for users of the cluster listing and working with CRDs.

Kubernetes operators are becoming the defacto standard for controlling infrastructure so many cluster admins will be interacting with CRDs on their clusters and it won't just be Crossplane deployed there. Having 2000 unused CRDs deployed to a cluster to operate Crossplane over multiple clouds when we only really need 30 of those CRDs is an undesirable user experience. It becomes much harder to navigate the list of CRDs and understand what is in use on a cluster vs what isn't. 

If Crossplane is going to be truly successful as a platform, personally I feel that it needs to be flexible enough to be used by teams who are subject to strict security and compliance controls and can't justify having 1000s of CRDs deployed without being in control of what they are, but equally as much, just those who like to run a really tight ship and don't like bloat. Even if the performance issues are fixed, the user experience issues remain unless a way to select or disable APIs is introduced.

To answer the question above about cross-resource-references - I would place responsibility for this in the hands of the users i.e. the teams developing the services on top of Crossplane. As long as the system fails with a clear error that API group B has been disabled so it cannot be used as a reference in group A then Crossplane users will easily identify and fix the issue without any harm done during the development phase of a service.

**Comment by negz** [comment]
> Does it stand to reason that retrieving a list of 1000s of resources using kubectl/k8s API is always going to be slower than a list of 10s of resources, no matter how much you fix upstream? So if a user ends up waiting longer and the majority of those resources are unnecessary then this is very frustrating for users of the cluster listing and working with CRDs.

I think perception of this problem is being exacerbated by the client-side rate limiting, which is an eminently fixable problem. The latest releases of kubectl are configured to burst to 300rps during discovery, which in [some initial tests](https://github.com/crossplane/crossplane/issues/2895#issuecomment-1162688419) indicate discovery of ~1,600 CRDs can complete in about a second. This makes sense given it's less than 10MB of text data - not too challenging for most modern internet connections. There's also ongoing work to make the process more efficient - it's really not at the moment. For example an [open KEP](https://github.com/kubernetes/enhancements/pull/3364) exists to consolidate discovery information into a single document that can better leverage etags so that discovery need only happen when something actually changes.

So with all this context, I don't think the "more will always be slower than less" argument is particularly compelling in the face of the additional cognitive, maintenance and implementation burden of allowing folks to granularly pick which CRDs a particular Crossplane package installs.

**Comment by carpenterm** [comment]
Thanks for the update on the performance work. This doesn't address the other issues around user experience, general bloat or highly regulated environments though. Not having the ability to choose the APIs a provider like Terrajet installs is going to result in extra work building custom providers for folks who care about this stuff or simply don't have a choice (e.g. https://github.com/crossplane/crossplane/issues/2869#issuecomment-1028239150) which essentially shifts the "cognitive, maintenance and implementation burden" onto all of those teams/users instead of handling it centrally in the tool itself on behalf of everyone. 

I know there are the reduced releases of Terrajet providers which provide less APIs, but I immediately ran into resources that were not supported by that version and had to downgrade to the older preview release which was when I started experiencing issues. Surely there is additional maintenance burden managing the preview releases separately to the major releases with handpicked APIs? Also additional cognitive load for maintainers and users having to manage what is included in those reduced releases? Having ability to control which groups of APIs got installed could remove the need to have those full and reduced releases.

Fixing performance issues upstream also means folks have to wait a long time to get a usable version. The ability to pick CRDs would enable users on older versions of Kubernetes who can't upgrade for other reasons outside of their control, say the latest versions of Kubernetes are not supported by their cloud provider or distribution and may not be for a long time. Research has shown that platform owners have a hard time keeping up to date with the latest releases and most businesses run several versions behind.

From a user perspective, the default experience and cognitive load need not change for those users who don't care about what gets installed, you could just install everything only selecting APIs if you run on older Kubernetes versions or are have company policies which require it, etc. I totally understand that there is an ongoing maintenance burden that comes with that, but that is a consideration for any feature and if it enables many more teams to use your product much sooner, hopefully that would be seen as a worthwhile investment.

Given that there are people in the community who are willing to work on this feature (e.g. https://github.com/crossplane/crossplane/pull/2646) is it fair to say that the real work is around coming up with an approach that the Crossplane maintainers are happy with that doesn't have the technical issues referenced here (https://github.com/crossplane/crossplane/issues/2869#issuecomment-1162660597) and minimizes the aforementioned maintenance burden as much as possible?

I hope these comments land in the right way, I'm purely trying to put across a user's perspective in a constructive way and provide additional information to consider to help you make a decision. Either way, it will be interesting to see how this issue progresses and if people keep asking for this - honestly I think Crossplane is a great tool from what I've seen so far so thank you for all your work on it, I just hope you reconsider this one in the future even if it turns out now isn't the right time.

**Comment by negz** [comment]
> This doesn't address the other issues around user experience, general bloat or highly regulated environments though.

I'd love to hear more detailed experience reports from folks in these highly regulated environments. What is it that the security teams are assessing about these CRDs? What do reviews look like? Are they doing code and/or schema audits of Crossplane providers, or is it more simple - "we want folks to be allowed to use S3 but not RDS" for example?

**Comment by slimm609** [comment]
@negz in the highly compliant environments, it is mixture of things. As you go to more and more restricted environments, this translates to stricter and stricter controls. When dealing with the U.S. DoD here is a reference of what controls/policies may apply to a project. https://csiac.org/wp-content/uploads/2022/04/2022-04-01-csiac-dod-cybersecurity-policy-chart.pdf. As you can see there are a ton of different groups of controls that can apply so it's very hard to give a specific answer to some of the concerns because it is highly dependent on who/what the customer does. 

- restricting what CRDs are needed for those specific environments instead of just a blank install everything. We do this today with 100+ other applications. We only install what is targeted for those specific clusters. This also allows for a better rollout of new features/resources. We feature flag what goes to a specific environment so we can control the rollout of certain functionality. 
- reducing the amount of data that needs to be explained and evaluated by the auditors. It follows the principle of least privilege and having to justify why resources supporting unauthorized or non-existent resources are even installed in the first place.
- Specifically with AWS, not all AWS partitions have service/feature parity.  Some of the AWS partitions only have a subset of the available services that exist in the standard commercial regions so installing a bunch of additional CRDs for services that don't exist and may never exist doesn't provide any benefit and have to answer to the auditors why it's even installed when the service doesn't exist in the first place. This is the same reason we don't install the GCP or Azure providers in AWS.  
- Not all environments have the connections that the "modern internet" does. Cloud providers deploy into completely isolated environments where connections are not the same as the standard internet and sometimes only a fraction of what they are in commercial cloud providers. 
- When dealing with clusters all around the world, 10MB of extra data per person per cluster can add up. 10MB from us-east-1 to someone on the east coast may not be much but 10MB from us-east-1 to someone in India or elsewhere can certainly add up. https://www.cloudping.co/grid shows that latency between some far away regions can get pretty high. 

the ability to restrict what CRDs are needed is a pretty important feature as @carpenterm said above. With supporting lots of different compliance requirements for many different environments we look at the compliance aspect before even really fully testing out the tool.  In certain cases like this where we can't restrict it then the answer may, unfortunately, be just to not use that application and find a different solution. If it was a single audit, it may be worth the explanation and effort but when doing 10+ independent audits every year and having to show year-over-year changes, it can quickly become not worth it.

**Comment by negz** [comment]
For those still experiencing performance issues due to rate limiting - here's the current state of the world to my knowledge.

The key issue here is that clients (like kubectl, helm, etc) need to occasionally perform a "discovery" process to understand what APIs (CRDs) the server supports. When there are too many CRDs this process takes a long time because clients are configured to limit the rate at which they query the server, and the more CRDs that are installed the more queries clients must perform.

So far this has primarily been addressed by https://github.com/kubernetes/kubernetes/pull/105520, which allows kubectl to make dramatically more queries to the server before any rate limiting kicks in. We've observed that this fix alone appears to eliminate the client-side issue for folks running the latest few versions of kubectl on Linux - at least 1.24.x and 1.23.9, possibly more.

There's a handful of further fixes we need to land, including:

* Fixing an additional slowdown on MacOS per https://github.com/kubernetes/kubernetes/issues/110753
* Rolling out the increased rate limits to all clients (e.g. Helm etc) per https://github.com/kubernetes/kubernetes/pull/109141

In addition to these changes that @ulucinar and I are managing upstream in Kubernetes, there's other promising work going on:

* Adding an experimental (alpha) discovery implementation that requires only a single client request per https://github.com/kubernetes/enhancements/pull/3364
* Discussion of removing discovery client rate limits all together.

If you happen to be using an M1 Mac, you're welcome to try [fastctl.zip](https://github.com/crossplane/crossplane/files/9012910/fastctl.zip). This is `kubectl` built from https://github.com/kubernetes/kubernetes/pull/110851. In our testing discovery with a cold cache typically takes about one second.

**Comment by mvaalexp** [comment]
Just FYI, we are now also seeing this which seems to be related to the number of CRDs installed
https://github.com/crossplane/crossplane/issues/2564

**Comment by ekristen** [comment]
Definitely would like to see the ability to not load all CRDs, it would be nice to be able to pick and choose. In my use case over 50% will never be used.

**Comment by dskatz** [comment]
We also have concerns about the additional CRDs and as previously stated, in our case as well over 50% will never be used.

**Comment by treyhyde** [comment]
A lot is said about clients above... remember that "crossplane" is itself a client and polls those CRDs constantly.  That then causes k8s audit events.  CRD audit events are fatter than most.  In EKS, those audit events are then sent to cloudwatch logs ... which are just about the most expensive thing on the planet.   In our simple setup, crossplane makes ~140k API requests, causing 2.6GB worth of k8s api audit logs to AWS per 30 min.... just to be "on"

**Comment by phyzical** [comment]
just thought i would add my usecase/issues relating to using crossplane in conjunction with argocd as its a blocker at this time to consider using it.

I can confirm terrajet/ the upbound aws-providers install fine and the crds are created.
But with these installed a sync takes upwards in the range of 7-10 minutes without any claims even created.
If i remove either of these providers and just use the contribute aws-provider (or no providers) everything is fine, so i think this issue related to the amount of crds.

incase it helps hint at any causes argocd runs in a seperate account and cluster to where crossplane is installed

* crossplane 1.10.1
* both clusters running eks 1.23

I assume has something to do with the amount of crds  given that the issue occurs with both terrajet and the unbound provider but is fine with the contribute aws provider

**Comment by iciclespider** [comment]
It seems the AWS provider is the main offender here.

What I think the AWS provider should do it break it into individual providers based on the aws api group. This is exactly what this AWS Java client does: https://github.com/aws/aws-sdk-java. Each AWS api group is its own jar, so only the api groups needed have to be part of the JVM classpath.

**Comment by Duologic** [comment]
> What I think the AWS provider should do it break it into individual providers based on the aws api group. 

This is actually a very nice idea that potentially could be easy to implement as a pattern in upjet/terrajet.

**Comment by yogeek** [comment]
> What I think the AWS provider should do it break it into individual providers based on the aws api group.

Isn't it the design chosen by AWS for their ACK controller ?
https://aws-controllers-k8s.github.io/community/docs/community/overview/#container-images

**Comment by ivanovaleksandar** [comment]
So, as many people pointed out the impact on user experience, this can be a deal breaker for those that are just starting out.  This issue becomes even more exaggerated when you want to monitor the deployed resources. You lack visibility because you are being met with a wall of CRDs that you don't know what they are.

But also looking at the practical side, deploy one resource and compare the performance of these two commands: `kubectl get <resource>` vs `kubectl get managed`. Obviously, the latter will feel like you are dragging a dead horse for no reason. Every time you want to see the status of your claim (a composed view of multiple related resources), Crossplane needs to go through all CRDs it deployed and check if there is a corresponding resource for it so it can give you the status. Two words about it: "no good".

So in summary, this makes me lean more towards an alternative (KCC, for example)

**Comment by eloo** [comment]
We have also discorded a negativ effect of this issue in combination with Kyverno. 
Kyverno is also trying to fetch the CRDs and thus puts some extra pressure on the system just because we are not able to limit the number of CRDs deployed by crossplane. 

The current approach really doesn't seem to scale.

**Comment by negz** [comment]
> What I think the AWS provider should do it break it into individual providers based on the aws api group

If we did take a path of allowing folks to install fewer CRDs (let's say allowing folks to optimize the ratio of installed to used CRDs) this would be my preferred approach as well, as opposed to allowing a configurable filter when installing a provider. Today's implementation of cross-resource references (e.g. `vpcIDRef`) wouldn't work under this model, so we'd need a good solution for https://github.com/crossplane/crossplane/issues/1770 first.

> So in summary, this makes me lean more towards an alternative (KCC, for example)

Won't KCC eventually have this problem as well? I believe they're at hundreds of CRDs. Obviously Crossplane is the extreme case here, but I wonder whether we're on a path to asking every project to let folks opt-out of the CRDs they don't need in order to workaround what seem to me issues in the scalability of the API server and its clients.

**Comment by julienduchesne** [comment]
For whoever it may help, the approach we're taking here is to fork the official providers and comment out the resources we don't need. It was pretty simple to do:

1. Commented out unneeded resources: https://github.com/grafana/crossplane-provider-azure/blob/main/config/externalname.go#L15
2. Commented out unneeded configs: https://github.com/grafana/crossplane-provider-azure/blob/main/config/provider.go#L115
3. `make generate`
4. Changed the publish account to our Upbound org: https://github.com/grafana/crossplane-provider-azure/blob/main/Makefile#L71
5. Push, and a version is created in our private provider
6. On a new release, we can cherry-pick our commit onto it

So, if it's possible to do in a static way, it should be possible to have support to select resources to install dynamically, right? I understand that references are the issue, but what if the framework resolved necessary dependencies when selecting a resource
Example: The user only selects PostgresSQL databases in the Azure provider, but these have a reference to Resource Groups, so the provider implicitly selects the resource groups as well to install

**Comment by mjnovice** [comment]
@julienduchesne I don't see `generate` as a target in the Makefile you have linked. Is there some other command we run ?

**Comment by mjnovice** [comment]
@julienduchesne @jvrplmlmn I tried doing it like that and it gave the below 

```
➜  crossplane-provider-azure git:(main) ✗ make generate
16:12:51 [ .. ] generating provider schema for hashicorp/azurerm 3.8.0
16:12:51 [ OK ] generating provider schema for hashicorp/azurerm 3.8.0
rm -fR "/Users/mayankjha/Code/crossplane-provider-azure/.work/terraform-provider-azurerm"
git clone -c advice.detachedHead=false --depth 1 --filter=blob:none --branch "v3.8.0" --sparse "https://github.com/hashicorp/terraform-provider-azurerm" "/Users/mayankjha/Code/crossplane-provider-azure/.work/terraform-provider-azurerm";
Cloning into '/Users/mayankjha/Code/crossplane-provider-azure/.work/terraform-provider-azurerm'...
remote: Enumerating objects: 1781, done.
remote: Counting objects: 100% (1781/1781), done.
remote: Compressing objects: 100% (1182/1182), done.
remote: Total 1781 (delta 17), reused 1214 (delta 9), pack-reused 0
Receiving objects: 100% (1781/1781), 513.88 KiB | 7.67 MiB/s, done.
Resolving deltas: 100% (17/17), done.
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 16 (delta 4), reused 4 (delta 1), pack-reused 0
Receiving objects: 100% (16/16), 176.45 KiB | 2.99 MiB/s, done.
Resolving deltas: 100% (4/4), done.
remote: Enumerating objects: 792, done.
remote: Counting objects: 100% (792/792), done.
remote: Compressing objects: 100% (567/567), done.
remote: Total 792 (delta 329), reused 318 (delta 225), pack-reused 0
Receiving objects: 100% (792/792), 962.29 KiB | 11.88 MiB/s, done.
Resolving deltas: 100% (329/329), done.
16:12:53 [ .. ] verify go modules dependencies have expected content
all modules verified
16:12:55 [ OK ] go modules dependencies verified
16:12:55 [ .. ] go generate linux_arm64

Generated 242 resources!
angryjet: error: error loading packages using pattern ./...: /Users/mayankjha/Code/crossplane-provider-azure/apis/cosmosdb/v1beta1/zz_generated.managed.go:11:11: undeclared name: Account
exit status 1
apis/generate.go:42: running "go": exit status 1
generate: open /Users/mayankjha/Code/crossplane-provider-azure/apis/cosmosdb/v1beta1/zz_account_types.go: no such file or directory
```
is there any other thing I need to cleanup ?

**Comment by lippertmarkus** [comment]
@mjnovice running `find . -iname 'zz_*' -delete` before `make generate` did the trick for me. You might also want to adapt the [platforms](https://github.com/grafana/crossplane-provider-azure/blob/main/Makefile#L15) before commiting and then run `make build` and `make publish`.

**Comment by uvegla** [comment]
This - implemented in some ways - feels like a much needed feature for us. Because of the large increase in CRDs we are experiencing a multitude of issues in our clusters. While increasing QPS / Burst on the discovery cache does have an impact it also surfaced a few new problems.

For example because of [Helm's lacking CRD management capabilities](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations) we have a custom mechanism that uses kubectl within a job to install / update / delete CRDs. The increase in burst means sending out ~300 request every time(without crossplane we barely had ~200 CRDs tops) because kubectl starts with an empty cache. This is well described [here](https://github.com/crossplane/crossplane/blob/master/design/one-pager-crd-scaling.md#client-side-throttling) (a really great doc, thanks for that!). With crossplane and hence ~1000 CRDs we observed very frequent OOM kills on these pods which in turn made our deployment very unreliable, often timing out by exhausting retries eventually even with 512M memory requests. To my knowledge the cache cannot be disabled.

We also client-side throttling issues in many of our controllers (and 3rd party tools as well, shared my thoughts on why it is not reliable in short / mid term [here](https://github.com/crossplane/crossplane/issues/3272#issuecomment-1424153259)) and bumping everything would take a considerable amount of effort or straight very difficult back-porting it old releases.

I guess it is heavily customer / use case dependent which CRDs are actually needed, but I would say in most cases 10-20% of the ~700 CRDs for the Upbound AWS provider are used in a large case scenario, quite possible less than 5% in general use cases. So having a convenient way to manage this feels like a very much needed feature from our side.

**Comment by negz** [comment]
I've removed this issue from the Crossplane Releases board and the 1.12 milestone. Rest assured though, that doesn't mean it's not an option we're considering. Having Crossplane providers install fewer CRDs is top priority for the project leads at the moment.

- https://github.com/crossplane/crossplane/issues/3852

I've created the above issue and added it to the project roadmap and the v1.12 milestone. This issue tracks the fact that we're working on assessing our options and producing a design to install fewer CRDs (without yet committing to a specific approach - e.g. filtering, smaller providers, etc). Do note that being in the v1.12 milestone means that leading up to v1.12 this is our top priority, but it's not a _guarantee_ that a fix will be available in v1.12. We'll do our best.

**Comment by nicolas-g** [comment]
As a temporary fix would a possible solution be to remove the CRDs immediately after installation?
```
kubectl delete crd crd_1
kubectl delete crd crd_2
kubectl delete crd crd_3
.
.
.
```

**Comment by negz** [comment]
> would a possible solution be to remove the CRDs immediately after installation

I don't think so - I haven't recently tested but I'm 99% sure you need to disable the relevant controllers too, or they'll crash the provider.

**Comment by negz** [comment]
After further investigation, breaking up the larger providers (https://github.com/crossplane/crossplane/issues/3754) continues to be my preferred alternative. I've opened a more detailed proposal at https://github.com/crossplane/crossplane/pull/3939. I'll put aside time at this week's community meeting on Thursday the 6th to discuss any concerns.

**Comment by mjnovice** [comment]
@negz that sounds great. Could you also share where can I join the community meeting ?

**Comment by negz** [comment]
@mjnovice sure - details are at https://github.com/crossplane/crossplane/#get-involved

**Comment by admssa** [comment]
b/c of the list of 901 resources installation of helm releases hangs if other crds are present in a chart. At least Karpenter installation hung on EKS through VPN until I removed all `upbound` resources. I really need to limit it somehow,  I need sqs and sns only.

**Comment by jbw976** [comment]
@admssa have you seen the just recently merged proposal to limit installed CRDS in #3939? We expect that be a big help here 💪

**Comment by negz** [comment]
Hi everyone!

We recently addressed the _performance_ part of this issue by breaking the largest providers up into families of smaller, more granular providers. You can read about that here:

* https://blog.crossplane.io/crd-scaling-provider-families/
* https://github.com/crossplane/crossplane/blob/2537e7d1/design/design-doc-smaller-providers.md

I understand that a lot of you raised concerns that were not related to performance in this issue. I've raised a new issue to track those:

* https://github.com/crossplane/crossplane/issues/4192

Please go 👍 that issue if it's something you want Crossplane to support.

In the meantime, I'm going to close and lock this issue. It's by far our most 👍 issue, but we're pretty confident we've fixed the main reason folks were up-voting it. We'd like to understand how many folks _specifically want filtering_, as opposed to wanting us to solve the performance issues they were facing, and the easiest way to do that is by creating a new issue.
